extensions/amp-user-location/0.1/test/test-amp-user-location.js

```diff
@@ -0,0 +1,264 @@
+        }
+        userLocation.appendChild(configElement);
+      }
+
+      return userLocation.build().then(() => userLocation);
+    }
+
+    it('should build if experiment is on', () => {
+      const element = newUserLocation();
+      return element.catch(unused => {
+        throw new Error('component should have built');
+      });
+    });
+
+    it('should not build if experiment is off', () => {
+      return allowConsoleError(() => {
+        toggleExperiment(env.win, 'amp-user-location', false);
+        return newUserLocation().catch(err => {
+          expect(err.message).to.include('experiment must be enabled');
+        });
+      });
+    });
+
+    it('should parse config when built, if present', () => {
+      return Promise.all([
+        newUserLocation()
+          .then(element => element.getImpl())
+          .then(impl => {
+            expect(impl.config_).to.be.null;
+          }),
+        newUserLocation({})
+          .then(element => element.getImpl())
+          .then(impl => {
+            expect(impl.config_).to.not.be.null;
+          }),
+      ]);
+    });
+
+    it('should error with invalid config', () => {
+      return allowConsoleError(() => {
+        return newUserLocation('this is not valid json').catch(err => {
+          expect(err.message).to.include(
+            'Failed to parse amp-user-location config'
+          );
+        });
+      });
+    });
+
+    it('should parse recognized fields in the config', () => {
+      const testConfig = {
+        fallback: '40,-22',
+        maxAge: 60000,
+        precision: 'low',
+        timeout: 10000,
+        doNotExpose: 'wow',
+      };
+
+      return newUserLocation(testConfig)
+        .then(element => {
+          return element.getImpl();
+        })
+        .then(impl => {
+          expect(impl.config_).to.include({
+            fallback: '40,-22',
+            maxAge: 60000,
+            timeout: 10000,
+          });
+          expect(impl.config_).to.not.have.property('doNotExpose');
+        });
+    });
+
+    it('should trigger the "approve" event if user approves geolocation', () => {
+      class UserLocationFake {}
+      UserLocationFake.prototype.requestLocation = sandbox.stub().resolves({
+        lat: 10,
+        lon: -10,
+      });
+      sandbox
+        .stub(Services, 'userLocationForDocOrNull')
+        .resolves(new UserLocationFake());
+
+      let triggerSpy;
+      return newUserLocation()
+        .then(element => element.getImpl())
+        .then(impl => {
+          triggerSpy = sandbox.spy(impl, 'triggerEvent_');
+          return impl.userLocationInteraction_();
+        })
+        .then(() => {
+          expect(triggerSpy).to.have.been.calledWith(
+            AmpUserLocationEvent.APPROVE,
+            {lat: 10, lon: -10}
+          );
+        });
+    });
+
+    it('should trigger the "deny" event if user denies geolocation', () => {
+      class UserLocationFake {}
```

Comment:
I like the declarative structure of stubs though, since it describes how the real method behaves in the situation. For example, the proposed `UserLocationFake#requestLocation` does not returns a `Promise` or reject and would behave differently than the real class.